### ISSUE TITLE
Implement suspendable RateLimiter

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#460]
Primitive, which distributes permits at the fixed rate or suspends current coroutine until it's available.
Such primitive is useful in a lot of server-side application to throttle requests to external services or other subsystem.

One of the first consumers of such feature may be Ktor (see https://github.com/ktorio/ktor/pull/473)

**Comment by vladimir-bukhtoyarov** [comment]
Bucket4j [already covers](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/AsyncScheduledBucket.java#L111) this functionality.

**Comment by tnleeuw** [comment]
> Bucket4j [already covers](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/AsyncScheduledBucket.java#L111) this functionality.

And so does Guava rate limiting, but neither does it with suspending functions and both seem to handle their blocking within their API, so you still have to wrap it.

**Comment by IRus** [comment]
> Primitive, which distributes permits at the fixed rate or suspends current coroutine until it's available.

Or immediately returns (like offer in channels), to respond on current request, that it was rate limited (for example).

**Comment by tnleeuw** [comment]
> > Primitive, which distributes permits at the fixed rate or suspends current coroutine until it's available.
> 
> Or immediately returns (like offer in channels), to respond on current request, that it was rate limited (for example).

If we want to implement rate-limiting in our own software, the most transparent way to do that is not to return until the ratelimit-interval has been reached.
Returning with a special code doesn't provide any convenience, because then you have to know when you can retry the operation -- the point of a suspending rate-limiter is to transparently handle this for you.

The situation here is not that you want to know that an API which you are calling returned a status that request-rate-limit was exceeded, but that internally you want to limit rate at which some events / requests / whatever are firing.

**Comment by IRus** [comment]
> If we want to implement rate-limiting in our own software, the most transparent way to do that is not to return until the ratelimit-interval has been reached.

Yes, it can be better in case when retry expected, but in case of overload (like you setup hard limit) you should drop request asap.

**Comment by tnleeuw** [comment]
> > If we want to implement rate-limiting in our own software, the most transparent way to do that is not to return until the ratelimit-interval has been reached.
> 
> Yes, it can be better in case when retry expected, but in case of overload (like you setup hard limit) you should drop request asap.

But this is not something which you can handle in a general-purpose rate limiter, because it depends on the actual call you make. The rate-limiter doesn't make the call, it only allows your co-routine to make calls at given max rate.
So in your own code, you should look at the return of whatever call you are making, to see if it makes sense to continue, retry, or abort.

**Comment by IRus** [comment]
> But this is not something which you can handle in a general-purpose rate limiter, because it depends on the actual call you make. The rate-limiter doesn't make the call, it only allows your co-routine to make calls at given max rate.

It depends on design of RateLimiter, so currently idk is it will be possible or not.

**Comment by czp3009** [comment]
```kotlin
val requestPerSecond = 20L
val rateLimiter = produce {
   while (true) {
        send(Unit)
        delay(1_000 / requestPerSecond)
    }
}
```

Use rateLimiter.receive() in the place needed.

Don't forget to cancel it.

**Comment by jensim** [comment]
I made a PR here #2799
And a tiny CI-pipe to test it [here](https://github.com/jensim/kotlinx.coroutines/actions/runs/985603272)

**Comment by jensim** [comment]
My PR now include both a RateLimiter and a IntervalLimiter, in case some more poeple are interested in taking a peek 🙂

Ive gotten tons of great feedback, and would appreciate more

**Comment by vlsi** [comment]
@jensim , hi, it is great you implemented a limiter, however, have you considered https://github.com/bucket4j/bucket4j?

It looks like they have well-thought API, and they cover many cases like bursts, etc.

**Comment by vlsi** [comment]
@qwwdfsad , I'm not sure what is your view, however, I just asked what bucket4j maintaners think of writing the core in Kotlin.

**Comment by vlsi** [comment]
@qwwdfsad , I'm not sure what is your view, however, I just asked what bucket4j maintaners think of writing the core in Kotlin.

**Comment by jensim** [comment]
That's a good suggestion @vlsi 

This would however bring rate limiting capabilities to all platforms without additional dependencies. And to me asynchronous execution often brings the need for setting a reasonable pace through some means. That's why I think it's a good thing to have a basic implementation available in the core. And if someone needs something more exact or has some other requirements, then other libs might satisfy the needs of that user on that platform.

**Comment by vlsi** [comment]
@jensim , my idea was to convince bucket4j to move to Kotlin, so bucket4j automatically becomes available on all the platforms.
In that regard, "minimal impl" does not sound to be that needed (especially, it would be sad to have slightly incompatible interfaces in-core and in the library)

The challenge is that the current bucket4j has no dependencies, which sounds like a plus for JVM consumers.
When migrating to Kotlin, it could probably avoid kotlin-stdlib, however, it would probably be hard to avoid `kotlinx-datetime` since client APIs should probably have something like `Duration` in the API.

**Comment by qwwdfsad** [comment]
It would be nice to have suspend capabilities in bucket4j even without multiplatform support. We probably can assist with reviewing coroutine-specific parts as well. I believe that they don't need to depend on `kotlinx-datetime` (i.e. `Duration` is part of the standard library) and cannot avoid standard library dependency anyway (e.g. the kotlinc puts enhanced null-checks on arguments of all public methods, and these checks are located in stdlib).

The biggest problem with getting rate limited in `kotlinx-coroutines` is its specificity and a huge space of possible configuration choices, which basically can be boiled down to two scenarios:

* We'll decide to develop a variety of operational modes, configurable parameters and extension points. In such a case, we are going to reinvent the whole bucket4j, and it's definitely not our goal
* We'll decide that we want to provide the simplest possible solution that suffices 90%+ of uses, something similar to Guava's rate limiter. It is much more realistic, but has its own problems: it requires a significant effort to research the design space and use-cases, and at this point, we can start questioning ourselves why it should be part of `kotlinx-coroutines` at all.

Anyway, it is out of our focus right now; IMO the best possible outcome is to have something like `bucket4j-kotlin` that is properly integrated with suspend API and recommend people using it

**Comment by qwwdfsad** [comment]
I don't think that shading standard library is a way to go; it may work for a while, but it definitely won't be future-proof

**Comment by jensim** [comment]
@vlsi 🤷🏼

My oppinion is that anything is better than nothing.

When I started looking at this, I hadn't found bucket4j, which would probably fit the use case I had back then. (A suspending rate limiter that would work with Kotlin coroutines, with very few dependencies)

In most JVM cases, a simple long var holding the last execution epoch time, and maybe a mutex, to delay any concurrent execution.. That neatly covers the RateLimiter in my oppinion. So I tend to reimplantat it every time, and tweak it to the needs of that case 🙈🥳 The interval limiter is more tricky, allowing a bucket of events per interval.

**Comment by ksletmoe** [comment]
Sorry to necro this thread, but if others found this because they were looking for a suspending / non-coroutine-blocking rate limiter (like I originally did), I wrote a Kotlin wrapper around bucket4j which suspends: https://github.com/ksletmoe/Bucket4k

It'd be great if the bucket4j folks rewrite the core in Kotlin, but in the mean time this should be useful.

**Comment by delder004** [comment]
I built a small library in this space that may be useful as a reference implementation: https://github.com/delder004/kotlin-rate-limiter

It’s a coroutine-native rate limiter for outbound/client-side throttling, with no runtime dependencies beyond `kotlinx-coroutines`.

  It currently includes:
  - bursty/token-bucket limiting
  - smooth/evenly paced limiting
  - composition for layered limits like `5/sec` + `100/min`

  A few design choices that worked well:
  - suspending `acquire()` plus non-suspending `tryAcquire()`
  - `Denied(retryAfter: Duration)` for immediate backoff decisions
  - lock-free `AtomicReference` + CAS state management, with lazy refill and no background coroutine
  - cancellation-safe `acquire()` that refunds reserved permits
  - injectable `TimeSource`, so tests work with `runTest` and virtual time

  Live demo: https://kotlin-rate-limiter-demo.dev/
